sdk/eventhub/azure-eventhub/tests/perfstress_tests/_test_base.py

```diff
@@ -3,37 +3,145 @@
 # --------------------------------------------------------------------------------------------
 
-from azure_devtools.perfstress_tests import PerfStressTest, get_random_bytes
+import asyncio
+from uuid import uuid4
+
+from azure_devtools.perfstress_tests import BatchPerfTest, EventPerfTest, get_random_bytes
 from azure.eventhub import EventHubProducerClient, EventHubConsumerClient, EventData
 from azure.eventhub.aio import (
     EventHubProducerClient as AsyncEventHubProducerClient,
     EventHubConsumerClient as AsyncEventHubConsumerClient
 )
 
 
-class _EventHubTest(PerfStressTest):
+class _EventHubProcessorTest(EventPerfTest):
     consumer_group = '$Default'
 
+    def __init__(self, arguments):
+        super().__init__(arguments)
+        connection_string = self.get_from_env("AZURE_EVENTHUB_CONNECTION_STRING")
+        eventhub_name = self.get_from_env("AZURE_EVENTHUB_NAME")
+
+        self.checkpoint_store = None
+        self.async_checkpoint_store = None
+        if arguments.use_storage_checkpoint:
+            storage_connection_str = self.get_from_env("AZURE_STORAGE_CONNECTION_STRING")
+            from azure.eventhub.extensions.checkpointstoreblob import BlobCheckpointStore
+            from azure.eventhub.extensions.checkpointstoreblobaio import BlobCheckpointStore as AsyncBlobCheckpointStore
+            from azure.eventhub.extensions.checkpointstoreblobaio._vendor.storage.blob import BlobServiceClient
+            self.container_name = "checkpointstore-" + str(uuid4())
+            self.blob_service_client = BlobServiceClient.from_connection_string(storage_connection_str)
+            self.checkpoint_store = BlobCheckpointStore.from_connection_string(storage_connection_str, self.container_name)
+            self.async_checkpoint_store = AsyncBlobCheckpointStore.from_connection_string(storage_connection_str, self.container_name)
+
+        self.consumer = EventHubConsumerClient.from_connection_string(
+            connection_string,
+            _EventHubProcessorTest.consumer_group,
+            eventhub_name=eventhub_name,
+            checkpoint_store=self.checkpoint_store,
+            load_balancing_strategy=arguments.load_balancing_strategy
+        )
+        self.async_consumer = AsyncEventHubConsumerClient.from_connection_string(
+            connection_string,
+            _EventHubProcessorTest.consumer_group,
+            eventhub_name=eventhub_name,
+            checkpoint_store=self.async_checkpoint_store,
+            load_balancing_strategy=arguments.load_balancing_strategy
+        )
+        if arguments.preload:
+            self.async_producer = AsyncEventHubProducerClient.from_connection_string(connection_string, eventhub_name=eventhub_name)
+
+    async def _preload_eventhub(self):
+        data = get_random_bytes(self.args.event_size)
+        async with self.async_producer as producer:
+            partitions = await producer.get_partition_ids()
+            for partition in partitions:
```

Comment:
Would it be faster to populate the partitions in parallel?

Reply:
A related question, why populate events partition by partition, rather than uploading to the gateway and letting the events be automatically partitioned?  Do we have scenarios where we want to measure direct access to partitions, as well as through the gateway?

Reply:
The reason that it's per partition is because I can only implement the 'idempotent message preloading' per partition - i.e. there's no way to detect cumulative messages in the EH. So I could count all the messages in each partition - then send the remainder to the gateway without prescribing partition - but I figured given that I was already iterating over the partitions - and when we count on the receiving end this is also done per partition - it seemed simpler just to load messages on that basis too.
Happy to update to the "count per partition but send to all partitions" model if you think it's preferable.

Reply:
Thanks for the explanation.  I do think it will be easier to use if we switch to "count per partition but send to all partitions", so can you please make this change?

Reply:
Yup - sure thing!

Reply:
Thanks @mikeharder - updated :)